### PR TITLE
Allow `falsy | T` spreads for falsy primitives

### LINKED ISSUE #16326: [Suggestion] Support conditionally setting keys with the spread operator
### Background
Support the behavior explained here: https://github.com/tc39/proposal-object-rest-spread/issues/45

This can be supported by extending `...undefined`, `...null` to also ignore `...boolean`.

### Example
There's already an example in the link above.
Personally, this is the snippet of code that I was trying to write:

```ts
this.headers = {
    ...this.headers,
    ...headers,
    ...authToken && { Authorization: authToken },
}
```

### Workaround
```ts
...(authToken ? { Authorization: authToken } : {})
```

### PR DESCRIPTION
**Pull request by sandersn**
Fixes #16326

 Allow falsy union types like `false | T` or `null | T` in spread in order to allow conditionally spreading:

```
function f(shouldSpread: boolean, origin: A, spreadee: B) {
  return { ...origin, ...shouldSpread && spreadee };
}
```

It's worth noting that this works currently without strictNullChecks because the type of `shouldSpread && spreadee` is B. However, with strictNullChecks it becomes `false | B`.

This PR special-cases changes the `isValidSpreadType` check to allow *falsy unions*, which are unions that have (1) have one or more falsy types (2) whose non-falsy parts return true for `isValidSpreadType`. This is actually *stricter* than before; these spreads are no longer allowed:

```ts
var n = { ...null };
var u = { ...undefined };
var un = { ...null, ...undefined };
```

Earlier in the PR's history, I had to to create a bespoke `Partial<T>` which was pretty clunky, but this code is not actually needed to fix the example from #16326, which I believe is by far the most common pattern.

### CODE REVIEW

**Comment by sandersn** [comment]
Actually, this needs to work for all falsy types. This means that all primitives are allowed in spread types now. Not sure I like this change, but I'll leave it open for discussion.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -7908,6 +7921,33 @@ namespace ts {
             return prop.flags & SymbolFlags.Method && find(prop.declarations, decl => isClassLike(decl.parent));
         }
 
+        function getPartialTypeFromFalsyUnion(type: UnionType): Type | undefined {
+            if (type.types.length === 2) {
+                // getFalsyFlagsOfTypes
+                // getTypeFacts
+                const i = Math.max(type.types.indexOf(falseType),
```
Comment (mhegazy) [inline src/compiler/checker.ts]: why not use `removeDefinitelyFalsyTypes` on the union, then check the identity of the output vs. input to know if you should get add `undefined` to the member when you are building the normal spread type.

Comment (sandersn) [reply src/compiler/checker.ts]: Good idea. That also covers undefined and null, which I forgot.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -7908,6 +7921,28 @@ namespace ts {
...
+            if (type.types.length === 2) {
+                const truthy = removeDefinitelyFalsyTypes(type);
+                if (truthy !== type) {
+                    const members = createSymbolTable();
+                    for (const prop of getPropertiesOfType(truthy)) {
+                        if (prop.flags & SymbolFlags.Optional) {
+                            members.set(prop.escapedName, prop);
+                        }
+                        else {
+                            const result = createSymbol(prop.flags | SymbolFlags.Optional, prop.escapedName);
```
Comment (mhegazy) [inline src/compiler/checker.ts]: can we merge this with the main spreadtype creation logic?

**Comment by sandersn** [comment]
I added the actual code from #16326 and confirmed that I do, in fact, need the code that creates the equivalent of `Partial<T>` when spreading the type `false | T` (or any falsy type, including `null | undefined`, etc). 

So I kept the code but made it more efficient and integrated with the existing `getSpreadType` code.

**Comment by sandersn** [comment]
OK, the ad-hoc `Partial<T>` equivalent code is gone now.

**Pull request merged**